ksql-streams/src/main/java/io/confluent/ksql/execution/streams/materialization/ks/KsLocator.java

```diff
@@ -51,27 +64,61 @@
   }
 
   @Override
-  @SuppressWarnings("deprecation")
-  public Optional<KsqlNode> locate(final Struct key) {
-    final StreamsMetadata metadata = kafkaStreams
-        .metadataForKey(stateStoreName, key, keySerializer);
-
-    if (metadata == StreamsMetadata.NOT_AVAILABLE) {
-      return Optional.empty();
+  public List<KsqlNode> locate(
+      final Struct key,
+      final Map<KsqlHost, HostStatus> allHostsStatus,
```

Comment:
One last pitch for the factory idea here.  You could have something like `List<RoutingFilter> routingFilters = factory.createRoutingFilters(activeHost, standByHosts);`

Within this factory, you could get allHostsStatus and pass it into the constructor of the filter, removing the need to pipe this all the way through.  Also, for lags, I could compute the max end offset from that same info and pass that into the constructor of the lag filter.  

Then we would modify a single factory which would register all of the filters and set them up.  Just an idea which would generalize fairly well.

Reply:
I can also do this in a followup with my lags change.